### DISCUSSION TITLE
Development tips

### DISCUSSION THREAD

**Discussion by Equalizer-0** [Disservin/fastchess#212] [Ideas]
Congratulations on the program, I have some suggestions for improving it.

I can't disable the creation of the .pgn file, I think that's the cause of the slowness at long times.
Is there a command to disable it?
I recommend disabling it by default, then if necessary you activate it with the -pgn command

On normal times the speed is the same as other programs, but on very fast times there are slowness problems.

I would also like to point out that the score is not indicated in the square brackets []

tc=0.2+0.002
Score of eng1 vs eng2: 9966 - 9643 - 391 [] 20000
fast-chess - time 16 min 43.44 sec - [0.508]

Score of eng1 vs eng2: 10510 - 8607 - 883  [0.548] 20000
c-chess-cli - time 4 min 34.75 sec

Score of eng1 vs eng2: 9837 - 9687 - 476  [0.504] 20000
cutechess - time  6 min 45.32 sec

**Comment by Disservin** [comment]
> I can't disable the creation of the .pgn file, I think that's the cause of the slowness at long times.

Thanks, it should now only try to write when the filename of the pgn option is specified. It might have slowed things down indeed, though I think cutechess might have a lot of timeouts which lead to shorter games, whereas in fastchess that shouldnt be the case and the games might take a bit longer.
Feel free to test it again.

**Reply by Equalizer-0** [reply]
Unfortunately with msvs I can't compile it, there are too many things to change, I will test the next executable that comes out, I also recommend the score in brackets.

**Reply by Disservin** [reply]
You don’t have to wait :) the ci uploads the executables after each push. https://github.com/Disservin/fast-chess/actions?query=event%3Apush+is%3Asuccess+branch%3Amaster
select the latest one and scroll to the bottom

**Reply by Equalizer-0** [reply]
Thanks, I'm really an amateur, I have a lot to learn.

**Reply by Equalizer-0** [reply]
In windows which compiler should I download to compile this program? MSYS2?

**Reply by Disservin** [reply]
Yeah I recommend following this setup 

https://code.visualstudio.com/docs/cpp/config-mingw#_installing-the-mingww64-toolchain

**Reply by Equalizer-0** [reply]
Excuse my inexperience, I installed everything, I run make -j and this comes out:
$ make -j
g++ -O3 -std=c++17 -Wall -Wextra -DNDEBUG -march=native -Isrc -Ithird_party -MMD -MP -c src/cli/cli.cpp -o tmp/src/cli/cli.o -static -static-libgcc -static-libstdc++ -Wl,--no-as-needed
g++ -O3 -std=c++17 -Wall -Wextra -DNDEBUG -march=native -Isrc -Ithird_party -MMD -MP -c src/engines/uci_engine.cpp -o tmp/src/engines/uci_engine.o -static -static-libgcc -static-libstdc++ -Wl,--no-as-needed
make: g++: No such file or directory
make: *** [Makefile:125: tmp/src/cli/cli.o] Error 127
make: *** Waiting for unfinished jobs....
make: g++: No such file or directory
make: *** [Makefile:125: tmp/src/engines/uci_engine.o] Error 127

**Reply by Disservin** [reply]
Did you do steps 4-7? You might have to restart your pc for the changes to take effect. `g++` seems to be not found currently.

**Reply by Equalizer-0** [reply]
Yes, I have done all the steps, including the verification.

D:\fast-chess>gcc --version
gcc (MinGW.org GCC-6.3.0-1) 6.3.0
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

D:\fast-chess>g++ --version
g++ (MinGW.org GCC-6.3.0-1) 6.3.0
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

D:\fast-chess>gdb --version
GNU gdb (GDB) 14.1
Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

---

If I run make -j from the administrator prompt it gives me these errors

D:\fast-chess>make -j
Access denied - SRC
The file could not be found - -NAME
Access denied - TESTS
The file could not be found - -MAXDEPTH
The file could not be found - 1
The file could not be found - -NAME
Access denied - SRC
The file could not be found - -NAME
Access denied - TESTS
The file could not be found - -MAXDEPTH
The file could not be found - 1
The file could not be found - -NAME
The specified path could not be found.
The specified path could not be found.
g++ -O3 -std=c++17 -Wall -Wextra -DNDEBUG -march=native -Isrc -Ithird_party -MMD -MP -o fast-chess  -static -static-libgcc -static-libstdc++ -Wl,--no-as-needed
c:/mingw/bin/../lib/gcc/mingw32/6.3.0/../../..\libmingw32.a(main.o):(.text.startup+0xa0): undefined reference to `WinMain@16'
collect2.exe: error: ld returned 1 exit status
make: *** [Makefile:122: fast-chess] Error 1

**Reply by Disservin** [reply]
Ah meh windows things, it uses the windows find over the find from mingw. One had to modify something in their path to get this right but I dont know at the moment :/

**Reply by Equalizer-0** [reply]
Thanks for your help, finally I managed to compile under Windows, I have to use the MSYS application in the UCRT64 version, it manages the environment variables in a different way.

**Reply by Disservin** [reply]
Nice :)

**Comment by Equalizer-0** [comment]
Score of eng1 vs eng2: 9854 - 9646 - 500 [0.51] 20000

time 17 min 1.19 sec

Great, now it doesn't print the pgn anymore.
The score is good, if you can put 3 numbers after the decimal point, so it becomes perfect.
The time did not decrease as I hoped by removing the pgn.
Strange, it's only noticeable on extremely short times, there are 33 milliseconds more each game, perhaps a wait time before the other game starts?

**Reply by Disservin** [reply]
what do you mean by `there are 33 milliseconds more each game,` ?

**Reply by Equalizer-0** [reply]
I use -concurrency 32, 32 games start, as soon as one game ends another one starts.
I mean the equivalent of cutechess -wait command
 -wait <n>
           Wait <n> milliseconds between games. The default is 0.

**Reply by Disservin** [reply]
well there is no hardcoded wait in the code

**Reply by Equalizer-0** [reply]
Then you have to optimize the code
